### DISCUSSION TITLE
Game compatibility issues with XInput HID Device Mode

### DISCUSSION THREAD

**Discussion by Kanuan** [nefarius/DsHidMini#114] [Support]
## Summary
In the new `XInput HID Device Mode` introduced in DshidMini v2.0.242, the controller acts as a DInput device with a XInput compatibility layer added on top of it. It works more or less as the following:

- Games that **don't** support generic controllers will recognize the DS3 as just an Xbox One controller
- Most games that do support both generic and XInput controllers should detect that both the DInput and Xbox controller are supposed to be one, therefore ignoring the DInput part of it (e.g.: Blasphemous)

The problem is that, some games, the controller:
- Will be detected as 2 separate devices (XInput and DInput parts recognized separately)
- Won't be detected at all
- Will be detected only as the xbox controller but with a broken layout
- Will appear to work as expected but then be unresponsive randomly
- Will only function as normal if Steam's `Xbox Configuration Support` is enabled

## Reproducing the issue
- Set the controller in `XInput HID Device Mode`
- Use the controller in one of the affected games

## Known affected games
See the ["Known afffected games" comment](https://github.com/nefarius/DsHidMini/discussions/114#discussioncomment-1047068) below

## Ways around the issue / Notes on what has been tested

(For users not waiting to Beta test the new mode, [a fully functional Xbox controller emulation can be achieved by using DS4Windows mode + software](https://docs.nefarius.at/projects/DsHidMini/DS4-Mode-User-Guide/))

- Games that suffer from issues with the beta XInput mode must be launched __through Steam__ and Steam's `Xbox Configuration Support` and `Generic Controller Configuration Support` must be enabled per-game or on Steam's General Controller Settings
    - The `Xbox Configuration Support` fixes games not detecting the XInput part of the controller
    - The `Generic Controller Configuration Support` prevents the game from detecting the DInput part of the controller

**Comment by Kanuan** [comment]
## Games that (apparently) do work as expected
- Dead cells
- Bloodstained: Curse of the moon 1 and 2
- Bloodstained: Symphony of the night
- Furi
- Resident Evil Remake
- Blasphemous
- Cuphead
- Baba is you
- Kentucky Route Zero
- Ultimate Chicken Horse
- The Witcher 3
- Firewatch
- The witness
- Portal 2
- Among Us
- Celeste

**Comment by Kanuan** [comment]
# Known affected games
## Hollow Knight
- Does not detect the controller on  default settings
- Using Steam's `Xbox Configuration Support` makes the game detect the controller but the controller does not fully work as expected, controller may also randomly stop working for a few seconds
- Necessary to activate the option `Use Native Input` while in-game then restart the game for the controller to work properly

## Dark Souls: Prepare to Die Edition
- Detect the controller but it does not work as expected (wrong button mapping)

## Devil May Cry 5
- On default settings game recognizes the controller but the input mappings are incorrect
- Work as normal with Steam's `Xbox Configuration Support` enabled

## Enter the Gungeon
- Controller randomly stops working
- Controller is detected as 2 separate devices, though this does not lead to issues since controllers can be manually defined and it defaults to the XInput device

How to reproduce (do not press or move anything that is not mentioned):
1) go in-game inside the gungeon
2) Hold L2 to open the map for 5s, then release the trigger
3) Move the Left stick in the upper-right direction for 2s then inmediatly to the lower-left direction in
**Note: I can't reliably reproduce this one. Because of this, this might be happening on other games too**.

## A short hike
- Game recognizes the controller only as a DInput device (icons show generic buttons instead of the xbox layout), though it works as expected without the need to rebind buttons

## Need For Speed Most Wanted (2005)
- Game recognizes the controller as 2 separate devices and then prompts that a unknown controller (the DInpit device) is not configured at every button press. Impossible to make the game ignore the second (DInput) controller, therefore making the game unplayable

## Tekken 7
- Game recognizes the controller as 2 separate devices
- Necessary to remap the DInput controller in-game so every button matches the XInput controller to be able to play the game in single player mode
- Impossible to play the game in 2 player mode

## Minecraft Windows 10 Version (not confirmed)
- Game does not detect the controller

## Yakuza 0
- Game recognizes the controller but with a broken layout

**Reply by Kanuan** [reply]
Hm... I just noticed that I'm doing all these tests via bluetooth. Will re-do them later on USB
## EDIT 1
Same behaviour on USB, tested on Hollow Knight and Dark Souls PTD

**Reply by Kanuan** [reply]
I've re-tested the games and having Steam's `Xbox Configuration Support` and `Generic Configuration Support`  options enabled at the same time, either per-game or on Steam's General Controller Settings, fixes the issues. The exception is Hollow Knight, which actually has an issue where Steam's `Xbox Config Support` breaks Xbox controller compatibility (RT/LT stop working) unless the "Native Controller support" in-game configuration is also enabled.

**Reply by tintingaroo** [reply]
Hollow Knight uses InControl. When HK is set with Native Controller Input (NCI) off, InControl is using "Unity Input" profiles where controllers are matched to a profile via device name e.g. "Controller (Xbox 360 For Windows)". "DS3 Compatible HID Device" is unknown to it.

With NCI on, InControl is using "Native Input" profiles, matching devices to profiles by (among other methods) VID/PID or driver type. Same as other cases, it sees the two devices:

Attached native device with handle 1:
Name: XInput Controller
Driver Type: XInput
Location ID: XInput::1
Serial Number: 
Vendor ID: 0xffff
Product ID: 0x0
Version Number: 0x1
Buttons: 14
Analogs: 6

Attached native device with handle 2:
Name: DS3 Compatible HID Device
Driver Type: DirectInput
Location ID: \\?\hid#vid_054c&pid_0268&ig_00#7&81f51f2&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}
Serial Number: TODO: use device address
Vendor ID: 0x7331
Product ID: 0x2
Version Number: 0x0
Buttons: 16
Analogs: 7

InControl will be detecting input from both. When it places the XInput device in the spot of Active Device, the game will respond to controls since there is an Xinput profile. When it sets the DInput device as the Active Device, there is loss of control as there is no profile for it, the VID/PID 7331/2 is unknown to it.

**Reply by Gibitwit** [reply]
Sacred 3 and Unreal Tournament 3 are affected.

Sacred 3 has the lower and upper sholder inputs swapped.

In Unreal Tournament 3 it is super busted with incorret mappings and the view constantly rotating to the left.  And constant move left input in the menus.

**Reply by nefarius** [reply]
@Gibitwit with what driver version did you test, this entire thread is quite old now.

**Reply by Gibitwit** [reply]
2.2.282.

**Reply by nefarius** [reply]
Well, there's your problem. Use v3, v2 is no longer in development and XI overhaul in v3 was the biggest change.

**Reply by Gibitwit** [reply]
Ok on the latest Pre relasese.

Doom Eternal thinks the gamepad is always sending input, thus constantly hides and recenters the mouse cursor.  Gamepad usage can be disabled for this game, but I did not need to do this with the SCP wrapper.

In Unreal Tournament 3 the dead zone overrides do not work.

**Comment by Kanuan** [comment]
> Nefarius' [comment](https://github.com/ViGEm/DsHidMini/issues/113#issuecomment-886161972): So TL;DR: what to do with this information?

I'm keeping this affected game list more as reference for future testing in case this mode receives bug-fixes. Because there are many different ways the incompatibility issue can manifest I though it would best to register it here so others testers (there are, like, 3 of us!) can benefit from it.

> Nefarius' [comment](https://github.com/ViGEm/DsHidMini/issues/113#issuecomment-886161972): The new approach may not be as universally perfect as simply emulating the good old x360 pad (which is impossible natively with DSHMs architecture) but it opens new ways to explore how XInput is processed in games and what to do to improve it
>  Nefarius' [comment](https://github.com/ViGEm/DsHidMini/issues/113#issuecomment-886160090): I was aware we're not gonna have a clean solution due to the sheer amount of shit game engines out there.

My biggest problem with this mode right now is calling it just "XInput", which gives the genaral idea of "_This mode should work with games that support xbox controllers_". You said once that you wanted to avoid a release with many asterisks attached, but because most of the issues described here can't be worked around and they happen in a significant number of games (as of now, in 7 of the 18 games tested) right now it is more like russian roulette when trying to play modern games.

Though DsHidMini v2 is in Beta, if some of the most pressing issues are not fixed (broken xbox layout on the games that do detect it, controller randomly stops responding for a few seconds on others) I think it would be better to rename this mode and make it recommended only to specific cases where it is known to work.

**Reply by nefarius** [reply]
I agree. We leave it in BETA for quite some time I recommend and see how screwed we really are 😁 After all I do admit I was confident rather quickly this time but then again, my small world, my small library of games, my rather individual use-case is satisfied so now comes "the world" educating us with what has been missed.

I still see it as a success though, like a month ago I wasn't even sure this was possible without another behemoth like ViGEmBus.

Plus, we didn't nuke DS4Windows compatibility, no even improved it, so a fallback to good old x360 pad emulation is available if all else fails.

**Reply by nefarius** [reply]
Also I was personally quite surprised how many games/engines do a combination of DI/HID/XI under the hood while simply claiming Xbox Controller compatibility. This is ofc. wrong and lazy marketing speech which bubbles up when you start API sniffing but for the general public User it's ofc a bad situation, it generates confusion. Welp, nobody said life was easy! 😁

**Reply by dxgldotorg** [reply]
Pretty sure back in the "Games for Windows" era that was technically against the rules for certification; an XInput controller had to use only XInput API.

**Comment by dxgldotorg** [comment]
Just created then upon finding this thread closed an Issue; when having trouble try forcing XBOX Controller support in the per-game settings in Steam.

**Reply by Kanuan** [reply]
Unfortunately it does not work on all games. Dark Souls PTD edition is one example where it remains broken. Need For Speed most wanted too.

### EDIT 1

Oh, you mean per-game config settings? Didn't try that, will take a look

**Reply by nefarius** [reply]
> Oh, you mean per-game config settings? Didn't try that, will take a look

Yes, that worked for me in Road Redemption.

**Reply by Kanuan** [reply]
I've re-tested the games and having Steam's `Xbox Configuration Support` and `Generic Configuration Support`  options enabled at the same time, either per-game or on Steam's General Controller Settings, fixes the issues. The exception is Hollow Knight, which actually has an issue where Steam's `Xbox Config Support` breaks Xbox controller compatibility (RT/LT stop working) unless the "Native Controller support" in-game configuration is also enabled.

**Comment by maryadavies** [comment]
I tried it with Final Fantasy XIV and no go, wouldn't even register the controller was there >.< I had to go back to DS4Windows.

All I can say is, not your fault, /punt Square Enix for having bad controller support. I'd be glad to help since I don't know if you got a lot of XIV players to test this with. But I swear, their pc games are notoriously finicky about controllers. *cough* FFXI *cough, cough*  I will try again with HidHide off if that could be the problem but I'd be glad to use my FFXIV account to help.

**Reply by Hardivh** [reply]
Way late, but maybe it can help lol
I managed to make it work by opening DSHMC.exe as administrator, and changing the HID Device Mode to "SXS (Steam. RCPS3)". I use a dualshock3 via bluetooth

**Reply by maryadavies** [reply]
Good to know tho for laymen who don't know, easier to use DS4Windows these days. Also I don't use steamed FFXIV (EWWWW! I know steaming XI caused some issues) so not sure if it'd work or not. Mine is the not-steam version.

**Comment by throwaway14** [comment]
You can include in the "Known affected games" list every game running on Godot. The Godot engine detects two gamepads instead of just one, though the engine is open source, it may help figuring out the issue.

**Reply by nefarius** [reply]
Is there a list/link to said games or the engine sources?

**Reply by throwaway14** [reply]
Godot Engine: https://github.com/godotengine/godot
Papope (game running on Godot Engine): https://github.com/atka66/papope

You may want to check AntiMicroX which doesn't show controller duplicates https://github.com/AntiMicroX/antimicrox

**Comment by AndnixSH** [comment]
I have a few games including an app that does not detect controller. I'm using PS3 controller
- Forza Horizon 3
- Forza Horizon 4
- Xbox app - To play cloud gaming (Workaround xbox.com/play)

Games that detect controller
- Forza Horizon 5

**Reply by Kanuan** [reply]
Unfortunately,  from what is known it is not possible to "fix" the XInput mode with the way it works currently, as it would require basically rebuilding DsHidMini from the ground up. Check the workarounds described on the 1st post for alternatives,

**Reply by AndnixSH** [reply]
I saw it, I was just letting you know the games I tried

**Comment by 503geek** [comment]
Default install of 2.2.282, set to XInput, DS3 on USB, not using Steam Controller Config. 

Confirmed, at least for me, that Minecraft Windows 10 doesn't see the controller. No failures otherwise.

**Tested working without trying rebinds:**
Worms W.M.D
Terraria
Stardew Valley
Final Fantasy XV
Streets of Rogue
Catherine Classic
Skyrim Special Edition

**Tested working with rebinds:**
Rocket League

If this is useful I can post more as I try them.

**Reply by nefarius** [reply]
We can not offer a solution other than the documented ones at the current moment in time but who knows what the future holds, be my guest to post more if you like!

**Comment by dxgldotorg** [comment]
Anyone have any luck with Sonic Generations?

**Comment by Leonidas903** [comment]
Hi, I wanted to tell that Fifa 23 has the same issue for counting double controller, nonetheless, as the years have passed, i haven't encountered any more problems with any other game I have played with my brother using DsHidMini, nor had the need to install other applications like DS4 windows, thanks for all your work

**Reply by Kanuan** [reply]
The main developer Nefarius made some changes to xinput mode and based on some tests the new xinput mode works perfectly.

It will be released along DsHidMini version 3, though there is no estimated beta release date for now

**Reply by Leonidas903** [reply]
That's great news! Thank you for the info 😄

**Comment by AnubyteCode** [comment]
Here's a fun, long workaround. 

Install this. Install DS4Windows. Run DS3 in SXS mode. Open DS4Windows in shared mode and it will detect and create a working xinput device. Next, switch the DS3 to run in xinput mode, DS4Windows will no longer see it, and games will now think it's a 360 controller. At least running games that previously had not. Huge pain but less driver stacking and no lingering HID hiding afaik.
